ci-bench/src/bench_lib/cachegrind.rs

```diff
@@ -0,0 +1,237 @@
+            Stdio::piped(),
+            Stdio::piped(),
+        )
+        .context("server side bench crashed")?;
+
+        let client = Self::run_bench_side(
+            &self.arch,
+            &self.executable,
+            benchmark_index,
+            Side::Client,
+            &bench.name_with_side(Side::Client),
+            Stdio::from(server.process.stdout.take().unwrap()),
+            Stdio::from(server.process.stdin.take().unwrap()),
+        )
+        .context("client side bench crashed")?;
+
+        let counts = InstructionCounts {
+            server: server.wait_and_get_instr_count()?,
+            client: client.wait_and_get_instr_count()?,
+        };
+
+        let overhead_counts = InstructionCounts {
+            server: self.overhead_instructions,
+            client: self.overhead_instructions,
+        };
+
+        Ok(counts - overhead_counts)
+    }
+
+    /// Returns the CPU architecture
+    fn get_arch() -> anyhow::Result<String> {
+        let output = Command::new("uname")
+            .arg("-m")
+            .stdout(Stdio::piped())
+            .output()
+            .context("Failed to run `uname` to determine CPU architecture.")?;
+
+        let arch = String::from_utf8(output.stdout)
+            .context("`-uname -m` returned invalid unicode.")?
+            .trim()
+            .to_owned();
+
+        Ok(arch)
+    }
+
+    /// Returns an error if cachegrind is not available
+    fn ensure_cachegrind_available() -> anyhow::Result<()> {
+        let result = Command::new("valgrind")
+            .arg("--tool=cachegrind")
+            .arg("--version")
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .status();
+
+        match result {
+            Err(e) => anyhow::bail!("Unexpected error while launching cachegrind. Error: {}", e),
+            Ok(status) => {
+                if status.success() {
+                    Ok(())
+                } else {
+                    anyhow::bail!("Failed to launch cachegrind. Error: {}. Please ensure that valgrind is installed and on the $PATH.", status)
+                }
+            }
+        }
+    }
+
+    /// See docs for [`Self::run_bench`]
+    fn run_bench_side(
+        arch: &str,
+        executable: &str,
+        benchmark_index: u32,
+        side: Side,
+        name: &str,
+        stdin: Stdio,
+        stdout: Stdio,
+    ) -> anyhow::Result<BenchSubprocess> {
+        let output_file = PathBuf::from(format!("target/cachegrind/cachegrind.out.{}", name));
+        std::fs::create_dir_all(output_file.parent().unwrap())
+            .context("Failed to create cachegrind output directory")?;
+
+        // Run under setarch to disable ASLR, to reduce noise
+        let mut cmd = Command::new("setarch");
+        let child = cmd
+            .arg(arch)
+            .arg("-R")
+            .arg("valgrind")
+            .arg("--tool=cachegrind")
+            .arg("--cache-sim=no")
```

Comment:
Maybe worth a comment similar to the `setarch` comment that mentions this disables collecting cache miss data? I had to read the fine manual to figure out what this arg does.

It also looks like this defaults to `no` (similar to `--branch-sim`). Is `cache-sim` important enough to be explicitly disabled, and if so, should `branch-sim` get the same treatment?

Reply:
Cache simulation is only disabled by default since the 3.21 release, which is pretty recent (e.g. Debian 12 ships valgrind 3.19). Branch prediction stats seem to be disabled by default already in 3.19, so probably don't need to be explicitly disabled. I'll add comments explaining the relevant arguments to valgrind.

Reply:
> Cache simulation is only disabled by default since the 3.21 release, which is pretty recent (e.g. Debian 12 ships valgrind 3.19)

Ahh! good call out. Thanks. I was looking at the upstream docs without any sense of what's shipped.

>  I'll add comments explaining the relevant arguments to valgrind. Do you think that would be enough here?

Yes, that sounds great. Thank you.